btcec/schnorr/musig2/context.go

```diff
@@ -548,6 +573,129 @@ func (s *Session) RegisterPubNonce(nonce [PubNonceSize]byte) (bool, error) {
 	return haveAllNonces, nil
 }
 
+// GenerateAdaptorSecret generates an adaptor secret key. This must be called
+// after the public nonces have been registered for all signers, because
+// the validity of the adaptor secret key depends on the combined nonce. It
+// must also be called before any partial signatures are generated or provided,
+// because the validation of the signatures depends on the adaptor point.
+func (s *Session) GenerateAdaptor(msg [32]byte) (*btcec.ModNScalar, *btcec.JacobianPoint, error) {
+	if s.combinedNonce == nil {
+		return nil, nil, ErrCombinedNonceUnavailable
+	}
+
+	if len(s.sigs) != 0 {
+		return nil, nil, ErrAlreadySigned
+	}
+
+	nonce, _, err := computeSigningNonce(
+		*s.combinedNonce, s.ctx.combinedKey.FinalKey, msg,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// The adaptor point added to the combined nonce must be even.
+	// We keep generating random secrets until we find one that works.
+	for {
+		var secretB [32]byte
+		_, err := rand.Read(secretB[:])
+		if err != nil {
+			return nil, nil, err
+		}
+
+		var secret btcec.ModNScalar
```

Comment:
This seems duplicated with `SetAdaptor`. Couldn't we instead just call into that method until we don't get `ErrInvalidAdaptorPoint`? Should reduce the size of the diff quite a bit.